### ISSUE TITLE
Knife SSH fails with signed ssh key

### ISSUE THREAD

**Issue by zarry** [chef/chef#5786]
## Description

Using Knife SSH with a signed ssh identify file fails.  We have a setup similar to what is outlined here: https://www.digitalocean.com/community/tutorials/how-to-create-an-ssh-ca-to-validate-hosts-and-clients-with-ubuntu.  We allow access to VMs through signing of SSH keys by a CA that is included on all boxes.  This functionality works fine for normal commands such as ssh <user>@<box>.  When attempting to use knife ssh and pass in the signed identify file it fails reverting back to user/pass.  

We believe this to be an original limitation on the net-ssh gem which knife ssh relies on.  This has since been fixed in net-ssh gem.  What are the chances we can get this functionality upstream into knife so that we can use commands such as knife bootstrap or knife ssh with a signed ssh key?  I believe there needs to be some changes in knife ssh to allow for this functionality but not 100% sure what the issue is. 

This was implemented in net-ssh in this commit: https://github.com/net-ssh/net-ssh/pull/485

## Chef Version

'''
╰─ chef --version
Chef Development Kit Version: 1.2.20
chef-client version: 12.18.31
delivery version: master (0b746cafed65a9ea1a79de3cc546e7922de9187c)
berks version: 5.5.0
kitchen version: 1.15.0
'''

## Platform Version

OSX 10.11.6

## Replication Case

You may need to follow a guide similar to what is outlined below. 

https://www.digitalocean.com/community/tutorials/how-to-create-an-ssh-ca-to-validate-hosts-and-clients-with-ubuntu

You should then be able to ssh to a given box using standard ssh with that signed key.  Using knife ssh will fail. 

## Client Output

I can provide this output but I had to muck around with a few things to even get the updated net-ssh gem working with knife so I am not sure how valuable that will be here. 

## Stacktrace

Likewise, I can provide a stacktrace but I was manually injecting the updated net-ssh gem into the chef gem space and forcing knife to use it.  This was not fruitful as I believe there are other compatibility issues with net-ssh 4.x and the current knife ssh code. 

Apologies for some of the vagueness here and I will gladly provide more information as needed.

**Comment by coderanger** [comment]
Can you link to the commit/PR where net-ssh added this? Would make it easier to track which version of things we need.

**Comment by zarry** [comment]
I updated the description to include the PR where this was fixed.  Thank you for the prompt response.

**Comment by coderanger** [comment]
There hasn't been a release of net-ssh since then, so there is nothing for us to do at this time. It doesn't look like this requires any code changes on our side so this will be picked up during the normal dependency upgrades in the ChefDK release after net-ssh releases the change. ChefDK releases are usually in the 4th week of each month.

**Comment by zarry** [comment]
@coderanger I agree with that assessment that there hasn't been an official release which includes this but I tried to validate it ahead of time by building the gem locally myself and forcing knife to use it.  

There was a breaking change on this line: https://github.com/net-ssh/net-ssh/blob/master/lib/net/ssh.rb#L263  as it used to default to true.  That is no longer the case and I initially received an argument error.  After tweaking that and getting knife to play nicely with this gem I still ran into an issue where it simply failed on using the key and reverted back to user/pass.  

I am not convinced this will just work when the latest version of chefdk pulls it in but I understand if there is not much action to take until that point.

**Comment by coderanger** [comment]
Unfortunately updating things piecemeal inside ChefDK is tricky because of how we package things. If you install the beta gem and then edit the top of your `knife` executable and update the `gem 'net-ssh'...` line it should probably work but you would be in somewhat uncharted territory since we haven't tested the beta of net-ssh :)

**Comment by zarry** [comment]
@coderanger That is essentially what I did but still ran into issues.  I did verify it was using the new version of the gem as in the -vv output of knife ssh I can see it using the version I specified.   I will prove out this functionality works in net-ssh gem next as maybe their implementation is flawed.  Then maybe look to follow back up for the knife side of things down the road when this makes it into an official release. 

I realize I am in uncharted waters here! I greatly appreciate the info and quick responses.

**Comment by zarry** [comment]
@coderanger We actually were able to get this working in our hacked together knife.  We ended up having to drop the -i and rely on ssh-agent to provide the keys.  When passing in a single key with -i set, it tripped the keys_only option which didn't allow the ssh agent to grab the other cert identity and pass that in.  At least thats my suspicion. 

Sounds like we just have to wait for an official release of net-ssh and then get that into knife. Appreciate the quick responses yesterday.  I believe this can likely be closed unless you want to keep it open until this gem version makes it into knife.
